File: Project4/sm3.py
import struct

IV = [
    0x7380166F,
    0x4914B2B9,
    0x172442D7,
    0xDA8A0600,
    0xA96F30BC,
    0x163138AA,
    0xE38DEE4D,
    0xB0FB0E4E,
]

T_j = [0x79CC4519] * 16 + [0x7A879D8A] * 48

def rotate_left(x, n):
    return ((x << n) & 0xFFFFFFFF) | (x >> (32 - n))

def FF_j(X, Y, Z, j):
    if 0 <= j <= 15:
        return X ^ Y ^ Z
    else:
        return (X & Y) | (X & Z) | (Y & Z)

def GG_j(X, Y, Z, j):
    if 0 <= j <= 15:
        return X ^ Y ^ Z
    else:
        return (X & Y) | (~X & Z)

def P_0(X):
    return X ^ rotate_left(X, 9) ^ rotate_left(X, 17)

def P_1(X):
    return X ^ rotate_left(X, 15) ^ rotate_left(X, 23)

def padding(msg):
    l = len(msg) * 8
    msg += b'\x80'
    while (len(msg) * 8) % 512 != 448:
        msg += b'\x00'
    msg += struct.pack('>Q', l)
    return msg

def msg_extend(B):
    W = []
    for i in range(16):
        W.append(struct.unpack('>I', B[i*4:i*4+4])[0])
    for j in range(16, 68):
        W.append(P_1(W[j-16] ^ W[j-9] ^ rotate_left(W[j-3], 15)) ^ rotate_left(W[j-13], 7) ^ W[j-6])
    W_1 = []
    for j in range(64):
        W_1.append(W[j] ^ W[j+4])
    return W, W_1

def CF(V_i, B_i):
    W, W_1 = msg_extend(B_i)
    A, B, C, D, E, F, G, H = V_i
    for j in range(64):
        SS1 = rotate_left((rotate_left(A, 12) + E + rotate_left(T_j[j], j)) & 0xFFFFFFFF, 7)
        SS2 = SS1 ^ rotate_left(A, 12)
        TT1 = (FF_j(A, B, C, j) + D + SS2 + W_1[j]) & 0xFFFFFFFF
        TT2 = (GG_j(E, F, G, j) + H + SS1 + W[j]) & 0xFFFFFFFF
        D = C
        C = rotate_left(B, 9)
        B = A
        A = TT1
        H = G
        G = rotate_left(F, 19)
        F = E
        E = P_0(TT2)
    V_j = [A ^ V_i[0], B ^ V_i[1], C ^ V_i[2], D ^ V_i[3],
           E ^ V_i[4], F ^ V_i[5], G ^ V_i[6], H ^ V_i[7]]
    return V_j

def sm3_hash(msg):
    msg = padding(msg)
    n = len(msg) // 64
    V = IV
    for i in range(n):
        B_i = msg[i*64:(i+1)*64]
        V = CF(V, B_i)
    result = b''.join(struct.pack('>I', x) for x in V)
    return result.hex()

File: Project4/sm3.cpp
#include "sm3.h"
#include <cstring>
#include <iostream>

// --- 常量定义 ---

static const uint32_t IV[8] = {
    0x7380166F, 0x4914B2B9, 0x172442D7, 0xDA8A0600,
    0xA96F30BC, 0x163138AA, 0xE38DEE4D, 0xB0FB0E4E
};

static inline uint32_t ROTL(uint32_t x, uint32_t n) {
    return (x << n) | (x >> (32 - n));
}

static inline uint32_t FF(uint32_t x, uint32_t y, uint32_t z, int j) {
    if (j <= 15) return x ^ y ^ z;
    else return (x & y) | (x & z) | (y & z);
}

static inline uint32_t GG(uint32_t x, uint32_t y, uint32_t z, int j) {
    if (j <= 15) return x ^ y ^ z;
    else return (x & y) | ((~x) & z);
}

static inline uint32_t P0(uint32_t x) {
    return x ^ ROTL(x, 9) ^ ROTL(x, 17);
}

static inline uint32_t P1(uint32_t x) {
    return x ^ ROTL(x, 15) ^ ROTL(x, 23);
}

static const uint32_t T_j[64] = {
    0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519,
    0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519, 0x79CC4519,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A,
    0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A, 0x7A879D8A
};

// --- 内部函数声明 ---

static void sm3_compress(SM3_CTX* ctx, const uint8_t block[64]);
static void sm3_padding(SM3_CTX* ctx, uint8_t out[SM3_DIGEST_LENGTH]);

// --- API 实现 ---

void sm3_init(SM3_CTX* ctx) {
    std::memcpy(ctx->state, IV, sizeof(IV));
    ctx->length = 0;
    ctx->buf_used = 0;
}

void sm3_update(SM3_CTX* ctx, const uint8_t* data, size_t len) {
    ctx->length += len * 8; // bit length
    while (len > 0) {
        size_t space = 64 - ctx->buf_used;
        size_t to_copy = (len < space) ? len : space;
        std::memcpy(ctx->buffer + ctx->buf_used, data, to_copy);
        ctx->buf_used += to_copy;
        data += to_copy;
        len -= to_copy;
        if (ctx->buf_used == 64) {
            sm3_compress(ctx, ctx->buffer);
            ctx->buf_used = 0;
        }
    }
}

void sm3_final(SM3_CTX* ctx, uint8_t out[SM3_DIGEST_LENGTH]) {
    sm3_padding(ctx, out);
}

std::vector<uint8_t> sm3_hash(const uint8_t* data, size_t len) {
    SM3_CTX ctx;
    uint8_t digest[SM3_DIGEST_LENGTH];
    sm3_init(&ctx);
    sm3_update(&ctx, data, len);
    sm3_final(&ctx, digest);
    return std::vector<uint8_t>(digest, digest + SM3_DIGEST_LENGTH);
}

// --- 核心压缩函数 ---

static void sm3_compress(SM3_CTX* ctx, const uint8_t block[64]) {
    uint32_t W[68], W_prime[64];
    // 消息扩展
    for (int i = 0; i < 16; ++i) {
        W[i] = (block[i * 4] << 24) | (block[i * 4 + 1] << 16) | (block[i * 4 + 2] << 8) | (block[i * 4 + 3]);
    }
    for (int j = 16; j < 68; ++j) {
        uint32_t tmp = W[j - 16] ^ W[j - 9] ^ ROTL(W[j - 3], 15);
        W[j] = P1(tmp) ^ ROTL(W[j - 13], 7) ^ W[j - 6];
    }
    for (int j = 0; j < 64; ++j) {
        W_prime[j] = W[j] ^ W[j + 4];
    }

    // 初始化寄存器
    uint32_t A = ctx->state[0];
    uint32_t B = ctx->state[1];
    uint32_t C = ctx->state[2];
    uint32_t D = ctx->state[3];
    uint32_t E = ctx->state[4];
    uint32_t F = ctx->state[5];
    uint32_t G = ctx->state[6];
    uint32_t H = ctx->state[7];

    for (int j = 0; j < 64; ++j) {
        uint32_t SS1 = ROTL((ROTL(A, 12) + E + ROTL(T_j[j], j)) & 0xFFFFFFFF, 7);
        uint32_t SS2 = SS1 ^ ROTL(A, 12);
        uint32_t TT1 = (FF(A, B, C, j) + D + SS2 + W_prime[j]) & 0xFFFFFFFF;
        uint32_t TT2 = (GG(E, F, G, j) + H + SS1 + W[j]) & 0xFFFFFFFF;
        D = C;
        C = ROTL(B, 9);
        B = A;
        A = TT1;
        H = G;
        G = ROTL(F, 19);
        F = E;
        E = P0(TT2);
    }

    ctx->state[0] ^= A;
    ctx->state[1] ^= B;
    ctx->state[2] ^= C;
    ctx->state[3] ^= D;
    ctx->state[4] ^= E;
    ctx->state[5] ^= F;
    ctx->state[6] ^= G;
    ctx->state[7] ^= H;
}

// --- 填充函数 ---

static void sm3_padding(SM3_CTX* ctx, uint8_t out[SM3_DIGEST_LENGTH]) {
    size_t i = ctx->buf_used;
    ctx->buffer[i++] = 0x80;
    if (i > 56) {
        while (i < 64) ctx->buffer[i++] = 0x00;
        sm3_compress(ctx, ctx->buffer);
        i = 0;
    }
    while (i < 56) ctx->buffer[i++] = 0x00;

    // 消息长度64bit大端写入
    uint64_t len_be = ctx->length;
    for (int j = 7; j >= 0; --j) {
        ctx->buffer[i++] = (len_be >> (8 * j)) & 0xFF;
    }
    sm3_compress(ctx, ctx->buffer);

    // 输出摘要
    for (i = 0; i < 8; ++i) {
        out[i * 4] = (ctx->state[i] >> 24) & 0xFF;
        out[i * 4 + 1] = (ctx->state[i] >> 16) & 0xFF;
        out[i * 4 + 2] = (ctx->state[i] >> 8) & 0xFF;
        out[i * 4 + 3] = ctx->state[i] & 0xFF;
    }
}
